server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -4644,30 +4644,34 @@ public void syncSegmentsFromRemoteSegmentStore(boolean overrideLocal, boolean re
                         remoteSegmentMetadata.getGeneration()
                     );
                     if (shouldCommit) {
-                        long processedLocalCheckpoint = Long.parseLong(infosSnapshot.getUserData().get(LOCAL_CHECKPOINT_KEY));
-                        // Following code block makes sure to use SegmentInfosSnapshot in the remote store if generation differs
-                        // with local filesystem. If local filesystem already has segments_N+2 and infosSnapshot has generation N,
-                        // after commit, there would be 2 files that would be created segments_N+1 and segments_N+2. With the
-                        // policy of preserving only the latest commit, we will delete segments_N+1 which in fact is the part of the latest
-                        // commit.
-                        Optional<String> localMaxSegmentInfos = localSegmentFiles.stream()
-                            .filter(file -> file.startsWith(IndexFileNames.SEGMENTS))
-                            .max(Comparator.comparingLong(SegmentInfos::generationFromSegmentsFileName));
-                        if (localMaxSegmentInfos.isPresent()
-                            && infosSnapshot.getGeneration() < SegmentInfos.generationFromSegmentsFileName(localMaxSegmentInfos.get())
-                                - 1) {
-                            // If remote translog is not enabled, local translog will be created with different UUID.
-                            // This fails in Store.trimUnsafeCommits() as translog UUID of checkpoint and SegmentInfos needs
-                            // to be same. Following code block make sure to have the same UUID.
-                            if (indexSettings.isRemoteTranslogStoreEnabled() == false) {
-                                SegmentInfos localSegmentInfos = store.readLastCommittedSegmentsInfo();
-                                Map<String, String> userData = new HashMap<>(infosSnapshot.getUserData());
-                                userData.put(TRANSLOG_UUID_KEY, localSegmentInfos.userData.get(TRANSLOG_UUID_KEY));
-                                infosSnapshot.setUserData(userData, false);
+                        // Replicas never need a local commit
+                        if (this.shardRouting.primary()) {
```

Comment:
Just an additional check in case it's missed on the caller end. Combined both of them into one if clause for readability.

Reply:
@Poojita-Raj : Sorry, if I was not clear enough in my last comment. What I mean here is to remove these nested if conditions. This can be done by moving the `shardRouting.primary()` check caller i.e. code places where [syncSegmentsFromRemoteSegmentStore](https://github.com/search?q=repo%3Aopensearch-project%2FOpenSearch+syncSegmentsFromRemoteSegmentStore&type=code) is invoked. The end result here is to have only one if condition.

Current code
```
if (shouldCommit) {
   if (routing.primary) {
   ...
   }
} else {
  ...
}
```

Proposal. This needs change in caller to add primary shard routing check. Example invocation would look like `syncSegmentsFromRemoteSegmentStore(false, true, shouldCommit && indexShard.routing.primary() == true);`
```
if (shouldCommit) {
  ...
} else {
  ...
}
```